test_suite/benchmark_test.py

```diff
@@ -0,0 +1,91 @@
+import sys
+import unittest
+import subprocess
+import os
+import glob
+
+import testenv
+
+class BenchmarkTest(unittest.TestCase):
+
+  class BenchmarkResults:
+    def __init__(self, title, resource_usage):
+      self.ru = resource_usage
+      self._title = title
+
+    @property
+    def title(self):
+      return self._title
+
+    @title.setter
+    def title(self, title):
+        self.title_ = title
+
+    @property
+    def time(self):
+      return "{0:.2f}".format((self.ru.ru_utime + self.ru.ru_stime) * 1000.0)
+
+    @property
+    def memory(self):
+      if sys.platform == 'darwin':
+        return "{0:.2f}".format(self.ru.ru_maxrss / 1024)
+      return "{0:.2f}".format(self.ru.ru_maxrss)
+
+
+  def benchmark(self, f, binary):
+    """Execute Starlark file, return stderr."""
+    proc = subprocess.Popen(
+        [glob.glob(testenv.STARLARK_BINARY_PATH[binary])[0], f],
+         stderr=subprocess.PIPE, stdout=subprocess.PIPE)
+
+    ru = os.wait4(proc.pid, 0)[2]
+    _, stderr = proc.communicate()
+    return self.BenchmarkResults(binary, ru), stderr
+
+  def check_output(self, output):
+    """Any line in stderr that doesn't start with ### is
+       considered an error and the test fails"""
+    for line in output.splitlines():
+      if not line.startswith('###'):
+        print("Unexpected: ", line.encode('utf-8'))
+        raise Exception("Test failed")
+    pass
+
+  def get_max_width(self, table, index):
+    """Get the maximum width of the given column index"""
+    return max([len(str(row[index])) for row in table])
+
+  def display_table(self, table):
+    """Pretty prints the results table"""
+    col_paddings = []
+    for i in range(len(table[0])):
+        col_paddings.append(self.get_max_width(table, i))
+
+    for row in table:
+        # left col
+        print(row[0].ljust(col_paddings[0] + 1), end='')
+        # rest of the cols
+        for i in range(1, len(row)):
+            col = row[i].rjust(col_paddings[i] + 2)
+            print(col, end='')
+        print()
+
+  def testFile(self):
+    print("===", test_file, "===")
+    f = os.path.join(testenv.STARLARK_TESTDATA_PATH, test_file)
+
+    results = []
+    for binary in testenv.STARLARK_BINARY_PATH:
+      result_, out = self.benchmark(f, binary)
```

Comment:
consider doing more than one run per binary to eliminate noise.

Reply:
for this you might find https://github.com/bazelbuild/bazel-bench/blob/master/utils/values.py useful /cc @joeleba